mlir/lib/Gradient/Transforms/GradMethods/HybridGradient.cpp

```diff
@@ -12,25 +12,383 @@
+            }
+            // We modify the callgraph in-place as we traverse it, so we can't use a symbol table
+            // here or else the lookup will be stale.
+            traverseCallGraph(clonedCallee, /*symbolTable=*/nullptr, [&](func::FuncOp funcOp) {
+                funcOp.walk([&](func::CallOp callOp) {
+                    if (callOp.getCallee() == qnode.getName()) {
+                        PatternRewriter::InsertionGuard insertionGuard(rewriter);
+                        rewriter.setInsertionPointToStart(&funcOp.getFunctionBody().front());
+                        Value paramCount =
+                            rewriter
+                                .create<func::CallOp>(loc, paramCountFn, callOp.getArgOperands())
+                                .getResult(0);
+                        callOp.setCallee(qnodeSplit.getName());
+                        callOp.getOperandsMutable().append(paramCount);
+                    }
+                });
+            });
+        }
+    }
+    return clonedCallee;
+}
+
+LogicalResult HybridGradientLowering::matchAndRewrite(GradOp op, PatternRewriter &rewriter) const
+{
+    if (op.getMethod() != "defer") {
+        return failure();
+    }
+
+    Location loc = op.getLoc();
+    func::FuncOp callee =
+        SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, op.getCalleeAttr());
+
+    SmallVector<Value> backpropArgs;
+    FailureOr<func::FuncOp> clonedCallee = cloneCallee(rewriter, op, callee, backpropArgs);
+
+    rewriter.setInsertionPoint(op);
+    std::vector<size_t> diffArgIndices = computeDiffArgIndices(op.getDiffArgIndices());
+    SmallVector<Value> backpropResults{op.getNumResults()};
+    // Iterate over the primal results
+    for (const auto &[cotangentIdx, primalResult] :
+         llvm::enumerate(clonedCallee->getResultTypes())) {
+        // There is one Jacobian per distinct differential argument.
+        SmallVector<Value> jacobians;
+        for (unsigned argIdx = 0; argIdx < diffArgIndices.size(); argIdx++) {
+            Type jacobianType =
+                op.getResultTypes()[argIdx * clonedCallee->getNumResults() + cotangentIdx];
+            if (isa<RankedTensorType>(jacobianType)) {
+                jacobians.push_back(rewriter.create<tensor::EmptyOp>(
+                    loc, jacobianType, /*dynamicSizes=*/ValueRange{}));
+            }
+            else {
+                jacobians.push_back(rewriter.create<arith::ConstantFloatOp>(
+                    loc, APFloat(0.0), cast<FloatType>(jacobianType)));
+            }
+        }
+
+        if (auto primalTensorResultType = dyn_cast<RankedTensorType>(primalResult)) {
+            // Loop over every entry of this result, creating a one-hot cotangent vector and running
+            // a backward pass via the BackpropOp.
+            iterateOverEntries(
+                primalTensorResultType, rewriter, loc,
+                [&, cotangentIdx = cotangentIdx](ValueRange indices) {
```

Comment:
> are we restricted to statically shaped tensors in the output then

Currently, yes. Lifting this restriction shouldn't be too complicated if that would be valuable to get in with this PR.

> would generating a loop around the backprop call allow us to support dynamically shaped outputs

Yes, this is how we'd do it.


> Could the code size become very large in the current implementation when the output tensor is large

Yes, generating a loop would be a way to prevent the code size from blowing up. However, with sufficiently many outputs, it would make sense to use something like forward mode rather than reverse mode. If we're just focused on the code size though, generating a loop would be the way to go.

Reply:
Thanks for the answers, I think we close this comment for now then :)